src/main/java/org/myrobotlab/math/MapperLinear.java

```diff
@@ -0,0 +1,237 @@
+      return minOutput;
+    }
+    if (maxOutput != null && c > maxOutput) {
+      log.warn("clipping output {} to {}", c, maxOutput);
+      return maxOutput;
+    }
+    return c;
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#calcOutputInt(int)
+   */
+  @Override
+  final public int calcOutputInt(int in) {
+    return (int) calcOutput(in);
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#calcOutputInt(double)
+   */
+  @Override
+  final public int calcOutputInt(double in) {
+    return (int) calcOutput(in);
+  }
+
+  public double getMaxX() {
+    return maxX;
+  }
+
+  public double getMaxY() {
+    return maxY;
+  }
+
+  public double getMinX() {
+    return minX;
+  }
+
+  public double getMinY() {
+    return minY;
+  }
+
+  public boolean isInverted() {
+    return inverted;
+  }
+
+  public void setInverted(boolean invert) {
+    // change state ONLY if needed
+    // we need only 1 logic : inverted=output inverted / output
+    // inverted=inverted
+
+    this.minOutput = minY;
+    this.maxOutput = maxY;
+    double t = minY;
+    if (invert) {
+      if (minY < maxY) {
+        minY = maxY;
+        maxY = t;
+      }
+      this.minOutput = maxY;
+      this.maxOutput = minY;
+    } else {
+      if (minY > maxY) {
+        minY = maxY;
+        maxY = t;
+      }
+      this.minOutput = minY;
+      this.maxOutput = maxY;
+    }
+
+    inverted = invert;
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#getMinOutput()
+   */
+  @Override
+  public double getMinOutput() {
+    return minOutput;
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#getMaxOutput()
+   */
+  @Override
+  public double getMaxOutput() {
+    return maxOutput;
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#setMinMaxOutput(double, double)
+   */
+  @Override
+  public void setMinMaxOutput(double minOutput, double maxOutput) {
+    this.minOutput = minOutput;
+    this.maxOutput = maxOutput;
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#getMinInput()
+   */
+  @Override
+  public double getMinInput() {
+    return minInput;
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#getMaxInput()
+   */
+  @Override
+  public double getMaxInput() {
+    return maxInput;
+  }
+
+  /* (non-Javadoc)
+   * @see org.myrobotlab.math.MapperInterface#setMinMaxInput(double, double)
+   */
+  @Override
+  public void setMinMaxInput(double minInput, double maxInput) {
+    this.minInput = minInput;
+    this.maxInput = maxInput;
+  }
+
+  public String toString() {
+    return String.format(" < i < %.2f => map(%.2f,%.2f,%.2f,%.2f) => %.2f < o < %.2f ", minInput, maxInput, minX, maxX, minY, maxY, minOutput, maxOutput);
+  }
+
+  public void map(double minX, double maxX, double minY, double maxY) {
+    if (maxY - minY == 0) {
+      log.error("maxY - minY == 0 ..  DIVIDE BY ZERO ERROR COMING !");

```

Comment:
This protects against divide by zero if this mapper has had map called on it.  (individually setting max/min values would skip this protection.

Reply:
Divide by zero should be protected where its calculated.
Additionally, the old mapper has "partially" separated clipping values minOutput & maxOutput but the minX and maxX values were overloaded to be clipping input :(

So, it should be one way or the other.  minY maxY should be used as clipping out values, or minX, maxX should have separate clipping input values.  The way it currently is in the original Mapper is least desirable.

I decided to have all separate
- min/max Input
- map values
- min/max Output

This allows the possibility of the relationship of ranges to be modified differently than "clip" values.